### DISCUSSION TITLE
Lit-HTML Injected Script Runs Twice

### DISCUSSION THREAD

**Discussion by James-Wilkinson-git** [storybookjs/storybook#28767] [Help] [answered]
### Summary

I wrote a decorator to inject a script tag to run functions for the web component and it works however it's been running twice on the story both in dev and prod modes.

```
import { html } from 'lit';

export const withScript = (scriptString: string) => (storyFn: () => any) => {
    const scriptTag = () => {
        const tag = document.createElement('script');
        tag.innerHTML =
            'document.addEventListener("DOMContentLoaded", function(event) {' +
            scriptString +
            '});';
        return tag;
    };
    return html`${storyFn()}${scriptTag()}`;
};
```
and the passed in script string runs twice :/ 

any ideas?

**Comment by James-Wilkinson-git** [comment]
Tried moving this out of a decorator and it also injects twice, tried on lit's website and it renders once, something in storybook is causing the lit-html template to mount twice which injects the script into the runtime twice.

**Comment by Marklb** [comment]
More information may be needed to know for sure. I don't know too much about details of lit-html, but a Storybook decorator should be expected to get called more than once. I don't remember what all can cause them to get called, but an arg changing would cause them to get called again.

Ideally the decorator should know how to handle itself on a second call. Without knowing what your scripts do, my first thought would be to have the script check property on the document, before it adds the event listener, and if the property exists then don't run the script and if it does not then set the property and run the script. Going a step further, you may be able to get a more accurate detection of when to run the script by listening to Storybook api events, but I would need to try and implement an example to be sure.

If you did my suggestion, you may have another problem that can be more tricky. If your script needs to run once per story, you would need to set a property for each story, but I don't remember a good way to do that at the moment. You can get the storyId from the context, but I think that id would be the same if you render the same story twice. For example, the docs page renders the first story twice.

Since I don't know a lot about lit-html, I may be over complicating it, but that is my thought based on what details you provided.

It is possible that your decorator is getting called more than it should, but even if that was fixed you will run into the problem again if you use controls.

**Comment by James-Wilkinson-git** [comment]
Yeah I've tried a few ways to make it only execute once the problem I am now having is that when you change pages its actually not reexecuting. When moving from a story to autodocs docs page it's not reexecuted. The scripts themselves also set const and then when reexecuted throws an error that it's already defined

**Comment by James-Wilkinson-git** [comment, marked as answer]
I have come up with the following solution to work with autodocs and storybook 8 using lit-html

Make a decorator to take in the script you want to run, and wrap that script in the DOMContentLoaded, this will work when moving between stories, but will not trigger if you move to an autodocs page.

```
export const withScript =
    (scriptString: string) => (storyFn: () => any, context: any) => {
        // Function to create a script element and inject it into the document
        function injectScript() {
            // Create and inject the new script
            const script = document.createElement('script');
            script.type = 'text/javascript';
            script.dataset.injected = 'custom-script'; // Mark the script as injected
            script.textContent = `
      document.addEventListener('DOMContentLoaded', function() {
        ${scriptString}
      }, false);
    `; // Wrap the script content in DOMContentLoaded event listener
            script.onload = () => {
                console.log('Script loaded successfully');
            };
            script.onerror = () => {
                console.error('Error loading script');
            };
            document.body.appendChild(script);
        }

        // Inject the script when the story renders
        injectScript();

        // Return the story to be rendered
        return storyFn();
    };
```
Now when we change pages we want to remove our injected scripts, and if we change to a /docs url which are the autodocs url force a refresh of the page so DOMContentLoaded can fire again. Added in preview-head.html

```
<script>
    // Initial URL
    let lastUrl = window.location.href;

    // MutationObserver to detect URL changes
    const observer = new MutationObserver(() => {
        if (lastUrl !== window.location.href) {
            // Remove any existing script with the same data attribute to prevent duplicates
            const existingScript = document.querySelector(
                'script[data-injected="custom-script"]',
            );
            if (existingScript) {
                console.log("Scripts Removed");
                existingScript.remove();
            }
            lastUrl = window.location.href;
            if (lastUrl.includes("docs")) {
                window.location.reload();
            }
        }
    });

    // Configuration for the observer
    const config = { subtree: true, childList: true };

    // Start observing the document
    observer.observe(document, config);
</script>
```
